File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public int goodNodes(TreeNode root) {
        return dfs(Integer.MIN_VALUE, root);
    }
    
    int dfs(int max, TreeNode root) {
        return root == null ? 0 : (root.val >= max ? 1 : 0) + dfs(Math.max(root.val, max), root.left) + dfs(Math.max(root.val, max), root.right);
    }
}

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
 public:
 void helper(TreeNode* root, int prev, int &count)
{
    if(root==NULL) return;
    if(root->val>=prev)       // if current node value is greater than prev value  update prev value and increase count
    {
        count++;
        prev= root->val;
    }
    helper(root->left,prev,count);
    helper(root->right,prev,count);
    
}
int goodNodes(TreeNode* root) {
    int count=0;
    helper(root,INT_MIN,count);
    return count;
    
}
};
